rust/agents/relayer/src/msg/gas_payment/mod.rs

```diff
@@ -18,72 +21,112 @@ mod policies;
+                    GasPaymentEnforcementPolicy::MeetsEstimatedCost { coingeckoapikey } => {
+                        Box::new(GasPaymentPolicyMeetsEstimatedCost::new(coingeckoapikey))
+                    }
+                    GasPaymentEnforcementPolicy::OnChainFeeQuoting { gasfraction } => {
+                        let gasfraction = gasfraction.replace(' ', "");
+                        let v: Vec<&str> = gasfraction.split('/').collect();
+                        assert_eq!(
+                            v.len(),
+                            2,
+                            r#"Could not parse gas fraction; expected "`numerator / denominator`""#
+                        );
+                        Box::new(GasPaymentPolicyOnChainFeeQuoting::new(
+                            v[0].parse::<u64>().expect("Invalid integer"),
+                            v[1].parse::<u64>().expect("Invalid integer"),
+                        ))
+                    }
+                };
+                (p, cfg.whitelist)
+            })
+            .collect();
 
-        Self {
-            policy,
-            whitelist,
-            db,
-        }
+        Self { policies, db }
     }
 }
 
 impl GasPaymentEnforcer {
-    /// Returns (gas payment requirement met, current payment according to the DB)
+    /// Returns Some(gas_limit) if the enforcer has approved the transaction or
+    /// None if the transaction is not approved.
     pub async fn message_meets_gas_payment_requirement(
         &self,
         message: &HyperlaneMessage,
         tx_cost_estimate: &TxCostEstimate,
-    ) -> Result<(bool, U256)> {
-        let current_payment = self.get_message_gas_payment(message.id())?;
+    ) -> Result<Option<U256>> {
+        let msg_id = message.id();
+        let current_payment = self.db.retrieve_gas_payment_for_message_id(msg_id)?;
+        let current_expenditure = self.db.retrieve_gas_expenditure_for_message_id(msg_id)?;
 
-        // If the message matches the whitelist, consider it as meeting the gas payment requirement
-        if self.whitelist.msg_matches(message, false) {
-            return Ok((true, current_payment));
+        for (policy, whitelist) in &self.policies {
+            if !whitelist.msg_matches(message, true) {
+                continue;
+            }
+            return policy
+                .message_meets_gas_payment_requirement(
+                    message,
+                    &current_payment,
+                    &current_expenditure,
+                    tx_cost_estimate,
+                )
+                .await;
         }
 
-        let meets_requirement = self
-            .policy
-            .message_meets_gas_payment_requirement(message, &current_payment, tx_cost_estimate)
-            .await?;
-
-        Ok((meets_requirement, current_payment))
+        panic!("No gas payment policy matched for message; consider adding a default policy to the end of the policies array which uses a wildcard whitelist. {message:?}")
```

Comment:
a little wary of panicking - wdyt of making this a warn log and just returning `Ok(None)`? I agree the relayer operator should be aware if this happens but panicking feels a bit intense

Reply:
I could make it an `error` level and I would be okay with that. I just think it needs to be loudly called out since it is fundamentally invalid configuration by how this is defined.

Reply:
happy with that